File: naive_bayes/cpp/src/nb_simple.cpp
#include <iostream>
#include <vector>
#include "../include/utils.h"
#include "../include/processesing.h"
#include <unordered_map>

using str_vec_t = std::vector<std::string>;
using map_str_vec_t = std::unordered_map<std::string, str_vec_t>;
using vec_msv_t = std::vector<map_str_vec_t>; // Vector of map_str_vec_t
using map_int_t = std::unordered_map<std::string, int>;




int main () {
  std::string projectRoot {getProjectRoot()}; // Get root path of project
  std::string filename {projectRoot + "/data/emails/email.csv"}; // filename = root + path to data
                                                                 //
  vec_msv_t  data; // Create map
  data = parseCSV(filename); // Read csv file into array<unordered_map<string, array<string>> 
                                                                          //
  vec_msv_t train_data, test_data; // Initialize train and test datasets
  std::tie(train_data, test_data) = train_test_split(data, 0.8); // Split data
  
  // Get probabilities of ham and spam in training.
  int len_ham {0};
    for (const auto& dict : train_data) {
        if (dict.find("ham") != dict.end()) {
            len_ham++;
        }
    }

  double ham_proba {static_cast<double>(len_ham) / train_data.size()};
  double spam_proba {1.0 - ham_proba};
  // Separate ham and spam data from training.
  vec_msv_t ham_data, spam_data;
  
  for (const auto& dict : train_data) {
    if (dict.find("ham") != dict.end()) {
      ham_data.push_back({{"ham", dict.at("ham")}});
    }
    if (dict.find("spam") != dict.end()) {
      spam_data.push_back({{"spam", dict.at("spam")}});
    }
  }

  // Create histograms for both ham and spam in training
  map_int_t ham_histogram, spam_histogram; // Create histogram map
  ham_histogram = build_histogram(ham_data);
  spam_histogram = build_histogram(spam_data);
  
  int num_ham_words {vocabulary_counter(ham_histogram)};
  int num_spam_words {vocabulary_counter(spam_histogram)};
      
  std::vector<std::string> y_hat, y;

  // Evaluation
  int alpha {1};
  for (const auto& email : test_data) {
    for (const auto& pair : email) {
      y.push_back(pair.first);
      double product_ham {compute_product(pair.second, ham_histogram, num_ham_words, ham_proba, alpha)};
      double product_spam {compute_product(pair.second, spam_histogram, num_spam_words, spam_proba, alpha)};
      y_hat.push_back(product_spam > product_ham ? "spam" : "ham");
    }
  }
  int correct {0};
  for (size_t i {0}; i < y_hat.size(); i++){
    if (y_hat[i] == y[i]) {
      correct++;
    }
  }
  std::cout << "Accuracy: " << (double) correct / test_data.size() << "%" << std::endl;
  return 0;
}

File: naive_bayes/python/nb_simple.py
import time
import sys
from pathlib import Path
sys.path.append(str(Path(__file__).resolve().parents[2]))

from utils.py_utils import build_dataset, train_test_split, clean_dataset
from config import BASE_DIR, DATA_PATH


start_time = time.time()


def build_historgram(dataset: list[dict[str, list[str]]])-> dict[str, int]:
    """ Builds a histogram, which is a dict with keys being unique words
    and values being the occurence of said word in that category (ham, spam)
    """
    word_dict: dict = {}
    for dict_ in dataset:
        for word_list in dict_.values():
            for word in word_list:
                if word not in word_dict:
                    word_dict[word] = 1 
                else:
                    word_dict[word] += 1 
    return word_dict

def compute_product(text: dict[str, list[str]], histogram: dict[str, int], num_words: int, proba: float, alpha: int = 1)-> float:
        """ The product of a document is computed by multiplying the histogram value for each term. Finally the 
        pronbability of a given class is also multiplied. 

        An alpha value is used incase a word is not present in a histogram, to be consistant, it is added to the value 
        if a word is found in a histogram.
            Ex: document 'The dog' = 'The' (0.8922)+alpha * 'dog' (0.034)+alpha * probability of ham (0.84) = 1.643489
        """
        product: float = 1.0 
        for word_list in text.values():
            for word in word_list:
                if word in histogram:
                    product *= ((histogram[word] + alpha ) / num_words)
                else:
                    product *= alpha / num_words
        return product * proba


path: Path = BASE_DIR / DATA_PATH / "emails" / "email.csv"
dataset: list[dict[str, str]] = build_dataset(path)
clean: list[dict[str, list[str]]] = clean_dataset(dataset)
train_data, test_data = train_test_split(clean, 0.8)  # Train test split_index

len_ham: int = sum([1 for dict in train_data if 'ham' in dict])

ham_proba: float = len_ham / len(train_data) 
spam_proba: float = 1.0 - ham_proba

# Make histograms 
spam_histo: dict[str, int] = build_historgram([dict for dict in train_data if 'spam' in dict])
ham_histo: dict[str, int] = build_historgram([dict for dict in train_data if 'ham' in dict])

# Total words in spam and ham
num_spam_words: int = sum(spam_histo.values())
num_ham_words: int = sum(ham_histo.values())
# Evaluation
results: list = []
for email in test_data:
    y_pred: str = list(email.keys())[0]
    product_ham: float = compute_product(email, ham_histo, num_ham_words, ham_proba)
    product_spam: float = compute_product(email, spam_histo, num_spam_words, spam_proba)
    results.append(("spam" if product_spam > product_ham else "ham", y_pred))
    
correct: int = 0 
for y_hat, y_pred in results:
    if y_hat == y_pred:
        correct += 1 

print(f"the accuracy is {correct / len(test_data)}")
end_time: float = time.time()
execution_time: float = end_time - start_time
print(f"Execution time: {execution_time} milliseconds")
